### PR TITLE
[2143] unique port names for multi-NIC VMs using network+subnet+vmname

### LINKED ISSUE #2143: Migrated VMs with multiple NICs get duplicate port names across subnets
## Issue

When a VMware VM with multiple network interfaces (NICs) attached to different subnets is migrated to PCD, all of the resulting OpenStack ports are named identically to the VM, with no differentiation between them.

OpenStack/PCD has no functional dependency on unique port names (resources are tracked by UUID internally), but some environments rely on human-readable, unique port names for operational identification and downstream automation.

Affected version: v0.4.6

## Expected outcome

Ports created for a VM with multiple interfaces should have unique, human-readable names after migration (e.g. `<VM_Name>-<interface_index>` or `<VM_Name>-<subnet_name>`).

## Suggested fix

- Update port naming logic (likely in `v2v-helper/openstack/openstack.go`, in the port/interface creation path) to append a unique identifier — interface index or subnet name — to the VM name instead of reusing the bare VM name for every port.
- Ensure the migration workflow passes interface index / subnet context through to the port creation function.

### PR DESCRIPTION
**Pull request by OmkarDeshpande7**
<html><head></head><body><h2>Summary</h2>
<p>Fixes #2143
When a VMware VM with multiple NICs is migrated, all OpenStack ports previously received identical names (<code>port-&lt;vmname&gt;</code>), making it impossible for operators and downstream automation to distinguish between them. A first pass fixed the cross-subnet case; a follow-up addresses the remaining case where two or more NICs land on the <em>same</em> subnet, which would still collide.</p>
<h2>Changes</h2>
<ul>
<li><strong>New <code>buildPortName</code> helper</strong> (<code>v2v-helper/pkg/utils/openstackopsutils.go</code>): constructs port names as <code>port-&lt;vmname&gt;-&lt;subnet&gt;-&lt;relative-index&gt;</code> (or <code>port-&lt;vmname&gt;-&lt;relative-index&gt;</code> for L2 networks, which have no subnet), enforcing the Neutron 255-char limit by truncating only the vmname segment. <code>relative-index</code> is a 1-based counter scoped to "this subnet (or no-subnet bucket) on this VM" — uniform, so even a VM's first NIC on a subnet gets <code>-1</code>, not a bare name.</li>
<li><strong>New <code>subnetPortIndex map[string]int</code> counter</strong>, threaded through <code>GetCreateOpts</code> and <code>ValidateAndCreatePort</code> (both signatures changed — <code>ValidateAndCreatePort</code> is an exported <code>OpenstackOperations</code> interface method, mock updated accordingly). <code>migrate.go</code>'s <code>createPortsForNetworks</code> creates one counter per VM and shares it across every NIC in that VM's port-creation loop, so a second NIC landing on the same subnet gets <code>-2</code> instead of colliding with the first.</li>
<li><strong>Updated <code>GetCreateOpts</code></strong> to resolve a subnet name (or fall back to the no-subnet bucket) and assign the indexed name across all four port-creation branches:
<ul>
<li>Normal (preserveIP=true, has IPs): uses first subnet matched to the VM's IP</li>
<li>No fixed IPs (preserveIP=false): performs an extra <code>network.Subnets[0]</code> lookup for subnet name</li>
<li>DHCP (nil IPs, network has subnets): performs a <code>network.Subnets[0]</code> lookup for subnet name (same mechanism as the no-fixed-IPs branch)</li>
<li>Nil IPs on an L2 network with zero subnets: no subnet to resolve, still gets an indexed no-subnet-bucket name; avoids indexing an empty subnet list</li>
</ul>
</li>
<li><strong>Unit tests</strong>: <code>TestBuildPortName</code> — 8 table-driven cases covering both subnet and L2 naming, first/second-NIC index increments, truncation edge cases, and the 255-char hard limit. New <code>TestGetCreateOpts_MultipleNICsSameSubnet_IndexIncrements</code> drives two <code>GetCreateOpts</code> calls sharing one counter map against a mock Neutron server and asserts distinct <code>-1</code>/<code>-2</code> port names for two NICs on the same subnet.</li>
</ul>
<h2>Port naming format</h2>

Case | Port name
-- | --
Normal / DHCP | port-\<vmname>-\<subnet-name>-\<relative-index>
L2 network | port-\<vmname>-\<relative-index>
vmname too long | port-\<vmname[:N]>-\<subnet-name>-\<relative-index> (truncated)

</body></html>

__Testing Done__

image

image

---

<a href="https://app.devin.ai/review/platform9/vjailbreak/pull/2147" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

### CODE REVIEW

**Comment by sarika-pf9** [comment]
@OmkarDeshpande7 this won't be unique if there is `two NICs, same VM, same network+subnet`

File: v2v-helper/pkg/utils/openstackopsutils.go [hunk 1]
```diff
@@ -735,6 +761,13 @@ func (osclient *OpenStackClients) GetCreateOpts(ctx context.Context, network *ne
 		// empty non-nil slice: user explicitly wants a port with no fixed IPs (preserveIP=false, no custom IP)
 		PrintLog("Creating port with no fixed IPs for mac " + mac)
 		createOpts.FixedIPs = []ports.IP{}
+		// Non-L2 networks have subnets; look up the first one for a meaningful port name.
+		if !isL2Network && len(network.Subnets) > 0 {
+			subnetID, err := subnets.Get(ctx, osclient.NetworkingClient, network.Subnets[0]).Extract()
```
Comment (sarika-pf9) [inline v2v-helper/pkg/utils/openstackopsutils.go:800]: Add PrintLog on error

**Review by spai-p9 — APPROVED** [review]

**Review by spai-p9 — APPROVED** [review]
looks good.
